File: ios/ReadApp/Services/MangaImageNormalizer.swift
import Foundation

enum MangaImageNormalizer {
    static func sanitizeUrlString(_ raw: String) -> String {
        var trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        
        // 1. 彻底纠正协议头畸形 (http// -> http://, https// -> https://, http:/ -> http:// 等)
        // 使用更稳健的正则表达式或前缀检查
        let lower = trimmed.lowercased()
        if lower.hasPrefix("http//") {
            trimmed = "http://" + trimmed.dropFirst(6)
        } else if lower.hasPrefix("https//") {
            trimmed = "https://" + trimmed.dropFirst(7)
        } else if lower.hasPrefix("http:/") && !lower.hasPrefix("http://") {
            trimmed = "http://" + trimmed.dropFirst(6)
        } else if lower.hasPrefix("https:/") && !lower.hasPrefix("https://") {
            trimmed = "https://" + trimmed.dropFirst(7)
        }
        
        // 再次检查 lower 状态，防止多次 drop 导致的逻辑错误
        let finalLower = trimmed.lowercased()
        if finalLower.hasPrefix("http://") || finalLower.hasPrefix("https://") {
            // 已经是规范的了
        } else if finalLower.contains("//") && !finalLower.contains("://") {
            // 处理类似 custom:// 但写成 custom// 的情况，或者 http// 漏掉了的情况
            if finalLower.hasPrefix("http") {
                trimmed = trimmed.replacingOccurrences(of: "//", with: "://", options: .anchored, range: nil)
            }
        }
        
        // 2. 移除常见的末尾逗号及其后的 Legado 额外参数 (如 ,{...})
        let patterns = ["\\.jpg", "\\.jpeg", "\\.png", "\\.webp", "\\.gif", "\\.bmp"]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]),
               let match = regex.firstMatch(in: trimmed, options: [], range: NSRange(location: 0, length: (trimmed as NSString).length)) {
                let end = match.range.location + match.range.length
                let length = (trimmed as NSString).length
                if end >= length { return trimmed }
                let prefix = (trimmed as NSString).substring(to: end)
                let suffix = (trimmed as NSString).substring(from: end)
                if suffix.hasPrefix(",") {
                    return prefix
                }
            }
        }
        if let idx = trimmed.range(of: ",%7B")?.lowerBound {
            return String(trimmed[..<idx])
        }
        if let idx = trimmed.range(of: ",{")?.lowerBound {
            return String(trimmed[..<idx])
        }
        return trimmed
    }

    static func normalizeHost(_ url: URL) -> URL {
        guard let host = url.host?.lowercased() else { return url }
        for rule in MangaImageNormalizationRules.hostRewrites {
            let from = rule.fromSuffix.lowercased()
            if host == from {
                return replaceHost(url, newHost: rule.toSuffix)
            }
            if host.hasSuffix("." + from) {
                let dropCount = from.count + 1
                let prefix = host.dropLast(dropCount)
                let newHost = prefix + "." + rule.toSuffix
                return replaceHost(url, newHost: String(newHost))
            }
        }
        return url
    }

    static func shouldPreferSignedUrls(results: [String], text: String) -> Bool {
        guard !results.isEmpty else { return false }
        guard text.contains("sign=") else { return false }
        let anySigned = results.contains { $0.contains("sign=") }
        if anySigned { return false }
        let allMatch = results.allSatisfy { url in
            guard let host = URL(string: url)?.host?.lowercased() else { return false }
            return MangaImageNormalizationRules.preferSignedHosts.contains { ruleHost in
                host == ruleHost || host.hasSuffix("." + ruleHost)
            }
        }
        return allMatch
    }

    private static func replaceHost(_ url: URL, newHost: String) -> URL {
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        components?.host = newHost
        return components?.url ?? url
    }
}

File: android/src/main/java/com/readapp/data/manga/MangaImageNormalizer.kt
package com.readapp.data.manga

import android.net.Uri

object MangaImageNormalizer {
    fun sanitizeUrlString(raw: String): String {
        var trimmed = raw.trim()
        val lower = trimmed.lowercase()
        if (lower.startsWith("http//")) {
            trimmed = "http://" + trimmed.removePrefix("http//")
        } else if (lower.startsWith("https//")) {
            trimmed = "https://" + trimmed.removePrefix("https//")
        } else if (lower.startsWith("http:/") && !lower.startsWith("http://")) {
            trimmed = "http://" + trimmed.removePrefix("http:/")
        } else if (lower.startsWith("https:/") && !lower.startsWith("https://")) {
            trimmed = "https://" + trimmed.removePrefix("https:/")
        }
        val patterns = listOf(".jpg", ".jpeg", ".png", ".webp", ".gif", ".bmp")
        val lower = trimmed.lowercase()
        for (pattern in patterns) {
            val idx = lower.indexOf(pattern)
            if (idx >= 0) {
                val end = idx + pattern.length
                if (end < trimmed.length && trimmed[end] == ',') {
                    return trimmed.substring(0, end)
                }
                return trimmed
            }
        }
        val idx1 = trimmed.indexOf(",%7B")
        if (idx1 >= 0) return trimmed.substring(0, idx1)
        val idx2 = trimmed.indexOf(",{")
        if (idx2 >= 0) return trimmed.substring(0, idx2)
        return trimmed
    }

    fun resolveUrl(raw: String, baseUrl: String): String {
        val cleaned = sanitizeUrlString(raw)
        val resolved = if (cleaned.startsWith("http")) {
            cleaned
        } else {
            if (cleaned.startsWith("/")) "$baseUrl$cleaned" else "$baseUrl/$cleaned"
        }
        return normalizeHost(resolved)
    }

    fun normalizeHost(url: String): String {
        val uri = Uri.parse(url)
        val host = uri.host?.lowercase() ?: return url
        for (rule in MangaImageNormalizationRules.hostRewrites) {
            val from = rule.fromSuffix.lowercase()
            if (host == from) {
                return uri.buildUpon().encodedAuthority(rule.toSuffix).build().toString()
            }
            if (host.endsWith(".$from")) {
                val prefix = host.substring(0, host.length - from.length - 1)
                val newHost = "$prefix.${rule.toSuffix}"
                return uri.buildUpon().encodedAuthority(newHost).build().toString()
            }
        }
        return url
    }

    fun normalizeCoverUrl(raw: String): String {
        val cleaned = sanitizeUrlString(raw)
        return if (cleaned.startsWith("http")) normalizeHost(cleaned) else cleaned
    }

    fun shouldPreferSignedUrls(results: List<String>, text: String): Boolean {
        if (results.isEmpty()) return false
        if (!text.contains("sign=")) return false
        if (results.any { it.contains("sign=") }) return false
        val allMatch = results.all { url ->
            val host = Uri.parse(url).host?.lowercase() ?: return@all false
            MangaImageNormalizationRules.preferSignedHosts.any { ruleHost ->
                host == ruleHost || host.endsWith(".$ruleHost")
            }
        }
        return allMatch
    }

    fun normalizeEscapedText(text: String): String {
        return text
            .replace("\\u002F", "/")
            .replace("\\/", "/")
            .replace("\\u003F", "?")
            .replace("\\u0026", "&")
    }
}
